android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/TesterInstrumentation.java

```diff
@@ -0,0 +1,96 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.android.integrationtest;
+
+import android.app.Instrumentation;
+import android.os.Bundle;
+import com.google.android.gms.common.GooglePlayServicesNotAvailableException;
+import com.google.android.gms.common.GooglePlayServicesRepairableException;
+import com.google.android.gms.security.ProviderInstaller;
+
+/**
+ * The instrumentation used to run the interop test from command line.
+ */
+public class TesterInstrumentation extends Instrumentation {
+  private String testCase;
+  private String host;
+  private int port;
+  private String serverHostOverride;
+  private boolean useTls;
+  private boolean useTestCa;
+
+  @Override
+  public void onCreate(Bundle args) {
+    super.onCreate(args);
+
+    testCase = args.getString("test_case", "empty_unary");
+    host = args.getString("server_host", "");
+    port = Integer.parseInt(args.getString("server_port", "0"));
+    serverHostOverride = args.getString("server_host_override", null);
+    useTls = Boolean.parseBoolean(args.getString("use_tls", "true"));
+    useTestCa = Boolean.parseBoolean(args.getString("use_test_ca", "false"));
+
+    if (useTestCa) {
+      InteropTester.initTestCa(getContext().getResources().openRawResource(R.raw.ca));
+    }
+
+    if (useTls) {
+      try {
+        ProviderInstaller.installIfNeeded(getContext());
+      } catch (GooglePlayServicesRepairableException e) {
+        // Bet the current security library is up to date or fail.
```

Comment:
For long term, we are going to support NPN fallback, so we still have chance to get TLS work when security provider installing fails.

Even in current situation, in some emulators, for example the Lollipop (api-level 21) without google api, the security provider installing would fail, but the ALPN still works.